### ISSUE TITLE
[Question]: Contrastive learning with multiple modalities

### ISSUE THREAD

**Issue by aleSuglia** [KevinMusgrave/pytorch-metric-learning#373]
Hi there, 

Thanks for putting together such a great library. I am interested in multimodal applications where contrastive learning is used to learn a shared embedding space where two embedding that belong to different modalities (e.g., text, image) are close together. A common loss is the InfoNCE loss where in-batch negatives are used. One possible implementation is the one used in CLIP: https://github.com/mlfoundations/open_clip/blob/main/src/training/train.py#L23

Any suggestions as to how to use this library to reproduce this kind of loss together with a `CrossBatchMemory`? I have the impression that the current implementation assumes you have a single embedding matrix. However,  I was thinking that by concatenating image and text embeddings, and then using labels to say which ones belong to the same "class", I should be able to achieve the same behaviour. Am I wrong?

Also, as far as I can see, the CrossBatchMemory reuses the labels from the previous iteration and doesn't compute new ones. For the self-supervised setup where the labels are just indexes, this won't work. Do I have to extend the class myself to take this into account?

Thanks,
Alessandro

**Comment by KevinMusgrave** [comment]
> Unless you're assuming that the `embeddings` associated to such labels are already the concatenation between image and text. Is that the case?

Yes

> EDIT:
> 
> Maybe something like this:
> 
> ```python
> import torch
> from pytorch_metric_learning.losses import NTXentLoss, CrossBatchMemory
> 
> 
> def create_labels(num_pos_pairs, previous_max_label):
>     # create labels that indicate what the positive pairs are
>     labels = torch.arange(0, num_pos_pairs)
>     labels = torch.cat((labels, labels)).to(previous_max_label.device)
>     # add an offset so that the labels do not overlap with any labels in the memory queue
>     labels += previous_max_label + 1
>     return labels
> 
> 
> contra_loss = NTXentLoss()
> cross_batch_memory = CrossBatchMemory(contra_loss, 128)
> 
> batches = [
>     (torch.randn(5, 128), torch.randn(5, 128)),
>     (torch.randn(5, 128), torch.randn(5, 128)),
>     (torch.randn(5, 128), torch.randn(5, 128)),
> ]
> 
> for vis_emb, text_emb in batches:
>     emb = torch.cat([vis_emb, text_emb])
>     labels = create_labels(vis_emb.shape[0], torch.max(cross_batch_memory.label_memory))
>     loss = cross_batch_memory(emb, labels)
>     print(loss)
> ```

I think that will work :+1:

**Comment by aleSuglia** [comment]
@KevinMusgrave I have the impression the loss is not going down though. I was checking the `combined_embeddings` and `combined_labels` and I found that the combined labels are these after 2 batches: 
```
tensor([5, 6, 5, 6, 1, 2, 1, 2, 3, 4, 3, 4, 5, 6, 5, 6])
```
But this makes me think that the memory is somehow wrong because it shouldn't contain the current batch (`torch.tensor([5,6,5,6])`). Any ideas?

**Comment by KevinMusgrave** [comment]
Have you tried ```NTXentLoss``` by itself?

**Comment by KevinMusgrave** [comment]
As long as positive pairs share the same label, I don't think there should be an issue. I would try ```NTXentLoss``` by itself as a sanity check. 

```CrossBatchMemory``` complicates things. The [paper](https://arxiv.org/pdf/1912.06798.pdf) suggests having a warmup period where memory isn't used so that the embeddings become more stable. So that is one possible issue.

**Comment by aleSuglia** [comment]
@KevinMusgrave it does indeed look correct now that I think about it. Could it be an issue with the reducer? Maybe I should be using a different reducer?
